### PR TITLE
[ANT-2300] Short-term storage: penalty for storage control, injection and withdrawal flow gradient

### PR DESCRIPTION
**Pull request by a-zakir**

### CODE REVIEW

File: src/solver/simulation/include/antares/solver/simulation/sim_structure_probleme_economique.h [hunk 1]
```diff
@@ -184,9 +190,11 @@ using AREA_INPUT = std::vector<::ShortTermStorage::PROPERTIES>; // index is loca
...
 {
     // Index is the number of the STS in the area
-    std::vector<double> level;      // MWh
-    std::vector<double> injection;  // MWh
-    std::vector<double> withdrawal; // MWh
+    std::vector<double> level;                   // MWh
+    std::vector<double> injection;               // MWh
+    std::vector<double> withdrawal;              // MWh
+    std::vector<double> costVariationInjection;  // MWh
+    std::vector<double> costVariationwithdrawal; // MWh
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/sim_structure_probleme_economique.h]: wrong unit : EUR/MWh

**Review by flomnes — COMMENTED** [review]
The new constraint and variables need to be instanciated only if the corresponding parameter `penalize-variation-withdrawal` or `penalize-variation-injection` is set to `true`.

**Comment by a-zakir** [comment]
> The new constraint and variables need to be instanciated only if the corresponding parameter `penalize-variation-withdrawal` or `penalize-variation-injection` is set to `true`.

while your were writing you comment, me i was writing the code 🙃

File: src/solver/simulation/include/antares/solver/simulation/sim_structure_probleme_economique.h (hunk 1 above)
Comment (a-zakir) [reply src/solver/simulation/include/antares/solver/simulation/sim_structure_probleme_economique.h]: thanks

File: src/libs/antares/study/include/antares/study/parts/short-term-storage/series.h [hunk 2]
```diff
@@ -26,9 +26,18 @@
...
+    bool injection;
+    bool withdrawal;
+};
+
 class Series
 {
 public:
+    explicit Series(const PenaltyCostOnVariation& penaltyCostOnVariation = PenaltyCostOnVariation{
+                      .injection = false,
+                      .withdrawal = false});
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/short-term-storage/series.h]: You can place these default values at l. 31 and 32 instead

File: src/libs/antares/study/include/antares/study/parts/short-term-storage/series.h [hunk 3]
```diff
@@ -47,6 +54,9 @@ class Series
     std::vector<double> costInjection;
     std::vector<double> costWithdrawal;
     std::vector<double> costLevel;
+    std::vector<double> costVariationInjection;
+    std::vector<double> costVariationWithdrawal;
+    PenaltyCostOnVariation penaltyCostOnVariation = {.injection = false, .withdrawal = false};
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/short-term-storage/series.h]:
```suggestion
    PenaltyCostOnVariation penaltyCostOnVariation;
```

File: src/libs/antares/study/include/antares/study/parts/short-term-storage/cluster.h [hunk 4]
```diff
@@ -47,5 +47,6 @@ class STStorageCluster
 
     std::shared_ptr<Series> series = std::make_shared<Series>();
     mutable Properties properties;
+    void SynchronizePenaltyOnVariation() const;
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/short-term-storage/cluster.h]: You probably don't need this synchronization mechanism.

However, you can move `struct PenaltyCostOnVariation` to `class Properties`, and remove anything related to penalties in `class Series`.

File: src/libs/antares/study/parts/short-term-storage/cluster.cpp [hunk 5]
```diff
@@ -78,11 +78,18 @@ bool STStorageCluster::validate() const
...
     series->fillDefaultSeriesIfEmpty(); // fill series if no file series
     return ret;
 }
 
+void STStorageCluster::SynchronizePenaltyOnVariation() const
+{
+    series->penaltyCostOnVariation = {.injection = properties.penalizeVariationInjection,
+                                      .withdrawal = properties.penalizeVariationWithdrawal};
+}
+
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/cluster.cpp]:
```suggestion
```

File: src/libs/antares/study/parts/short-term-storage/series.cpp [hunk 6]
```diff
@@ -49,6 +49,15 @@ bool Series::loadFromFolder(const fs::path& folder)
     ret = loadFile(folder / "cost-withdrawal.txt", costWithdrawal) && ret;
     ret = loadFile(folder / "cost-level.txt", costLevel) && ret;
 
+    if (penaltyCostOnVariation.injection)
+    {
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/series.cpp]:
```suggestion
```
`loadFile` tolerates the absence of files

File: src/libs/antares/study/parts/short-term-storage/series.cpp [hunk 7]
```diff
@@ -122,6 +131,16 @@ void Series::fillDefaultSeriesIfEmpty()
     fillIfEmpty(costInjection, 0.0);
     fillIfEmpty(costWithdrawal, 0.0);
     fillIfEmpty(costLevel, 0.0);
+
+    if (penaltyCostOnVariation.injection)
+    {
+        fillIfEmpty(costVariationInjection, 0.0);
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/series.cpp]:
```suggestion
        fillIfEmpty(costVariationInjection, 0.0);
```

File: src/libs/antares/study/parts/short-term-storage/series.cpp [hunk 8]
```diff
@@ -148,6 +167,15 @@ bool Series::saveToFolder(const std::string& folder) const
     checkWrite("cost-withdrawal.txt", costWithdrawal);
     checkWrite("cost-level.txt", costLevel);
 
+    if (penaltyCostOnVariation.injection)
+    {
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/series.cpp]:
```suggestion
```

File: src/solver/optimisation/constraints/ShortTermStorageCostVariationInjectionBackward.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    const int hourInTheYear = builder.data.weekInTheYear * 168 + pdt;
+    namer.UpdateTimeStep(hourInTheYear);
+    namer.UpdateArea(builder.data.NomsDesPays[pays]);
+
+    builder.updateHourWithinWeek(pdt);
+    for (const auto& storage: data.ShortTermStorage[pays])
+    {
+        if (storage.penalizeVariationInjection)
+        {
+            namer.ShortTermStorageCostVariationInjectionBackward(builder.data.nombreDeContraintes,
```
Comment (flomnes) [inline src/solver/optimisation/constraints/ShortTermStorageCostVariationInjectionBackward.cpp]:
```suggestion
            namer.shortTermStorageCostVariationInjectionBackward(builder.data.nombreDeContraintes,
```

**Review by flomnes — COMMENTED** [review]
- Factorize the constraint builder classes, possibily using virtual classes or templates
- You don't need `Series` to know about parameters

File: src/solver/optimisation/include/antares/solver/optimisation/constraints/ShortTermStorageCostVariation.h [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * You should have received a copy of the Mozilla Public Licence 2.0
+ * along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+ */
+
+#pragma once
+#include "ConstraintBuilder.h"
+
+template<typename BuilderFunc, class PenaltyBoolean, class ConstraintIndex>
+void addStorageConstraint(BuilderFunc buildConstraint,
```
Comment (flomnes) [inline src/solver/optimisation/include/antares/solver/optimisation/constraints/ShortTermStorageCostVariation.h]: `class PenaltyBoolean` is actually always `bool`, so you can remove it

**Review by flomnes — COMMENTED** [review]
Please update the [migration guide](https://antares-simulator.readthedocs.io/en/latest/user-guide/04-migration-guides)

File: src/solver/optimisation/include/antares/solver/optimisation/constraints/ShortTermStorageCostVariation.h (hunk 10 above)
Comment (a-zakir) [reply src/solver/optimisation/include/antares/solver/optimisation/constraints/ShortTermStorageCostVariation.h]: true or false? depends on s-t object

Comment (a-zakir) [reply src/solver/optimisation/include/antares/solver/optimisation/constraints/ShortTermStorageCostVariation.h]: .. to decide if we do injection or withdrawal

Comment (flomnes) [reply src/solver/optimisation/include/antares/solver/optimisation/constraints/ShortTermStorageCostVariation.h]: Same as other parameters, their type is always the same. They can be removed

Comment (a-zakir) [reply src/solver/optimisation/include/antares/solver/optimisation/constraints/ShortTermStorageCostVariation.h]: how ? the loop on sts is inside addStorageConstraint

File: src/solver/optimisation/opt_gestion_second_membre_cas_lineaire.cpp [hunk 11]
```diff
@@ -36,9 +36,31 @@ static void shortTermStorageLevelsRHS(
...
+
+            // may be not needed if SecondMembre is 0 initialiazed
+            if (storage.penalizeVariationWithdrawal)
+            {
+                cnt = CorrespondanceCntNativesCntOptim
+                        .ShortTermStorageCostVariationWithdrawalForward[clusterGlobalIndex];
+                SecondMembre[cnt] = 0.;
+                cnt = CorrespondanceCntNativesCntOptim
+                        .ShortTermStorageCostVariationWithdrawalBackward[clusterGlobalIndex];
+                SecondMembre[cnt] = 0.;
```
Comment (flomnes) [inline src/solver/optimisation/opt_gestion_second_membre_cas_lineaire.cpp]: `SecondMembre` is initialized to 0, so these can be removed safely

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 12]
```diff
@@ -191,7 +191,33 @@ static void setBoundsForShortTermStorage(PROBLEME_HEBDO* problemeHebdo,
...
+                // is this necessary?
+                if (storage.penalizeVariationWithdrawal)
+                {
+                    int varCostVariationWithdrawal = variableManager
+                                                       .ShortTermStorageCostVariationWithdrawal(
+                                                         clusterGlobalIndex,
+                                                         pdtJour);
+                    Xmin[varCostVariationWithdrawal] = 0.;
+                    AddressForVars[varCostVariationWithdrawal] = &STSResult.costVariationwithdrawal
+                                                                    [storageIndex];
```
Comment (flomnes) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]:
```suggestion
```

File: src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp [hunk 13]
```diff
@@ -191,7 +191,33 @@ static void setBoundsForShortTermStorage(PROBLEME_HEBDO* problemeHebdo,
...
+                {
+                    int varCostVariationInjection = variableManager
+                                                      .ShortTermStorageCostVariationInjection(
+                                                        clusterGlobalIndex,
+                                                        pdtJour);
+
+                    Xmin[varCostVariationInjection] = 0.;
+
+                    AddressForVars[varCostVariationInjection] = &STSResult.costVariationInjection
+                                                                   [storageIndex];
```
Comment (flomnes) [inline src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp]:
```suggestion
```

File: src/solver/simulation/sim_alloc_probleme_hebdo.cpp [hunk 14]
```diff
@@ -433,6 +447,10 @@ void SIM_AllocateAreas(PROBLEME_HEBDO& problem,
             problem.ResultatsHoraires[k].ShortTermStorage[pdt].withdrawal.resize(
               nbShortTermStorage);
             problem.ResultatsHoraires[k].ShortTermStorage[pdt].level.resize(nbShortTermStorage);
+            problem.ResultatsHoraires[k].ShortTermStorage[pdt].costVariationInjection.resize(
+              nbShortTermStorage);
+            problem.ResultatsHoraires[k].ShortTermStorage[pdt].costVariationwithdrawal.resize(
+              nbShortTermStorage);
```
Comment (flomnes) [inline src/solver/simulation/sim_alloc_probleme_hebdo.cpp]:
```suggestion
```

File: src/solver/simulation/include/antares/solver/simulation/sim_structure_probleme_economique.h [hunk 15]
```diff
@@ -184,9 +192,11 @@ using AREA_INPUT = std::vector<::ShortTermStorage::PROPERTIES>; // index is loca
...
 {
     // Index is the number of the STS in the area
-    std::vector<double> level;      // MWh
-    std::vector<double> injection;  // MWh
-    std::vector<double> withdrawal; // MWh
+    std::vector<double> level;                   // MWh
+    std::vector<double> injection;               // MWh
+    std::vector<double> withdrawal;              // MWh
+    std::vector<double> costVariationInjection;  // EUR/MWh
+    std::vector<double> costVariationwithdrawal; // EUR/MWh
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/sim_structure_probleme_economique.h]:
```suggestion
```

**Review by flomnes — COMMENTED** [review]
Shorter

**Review by flomnes — APPROVED** [review]

**Pull request merged**